coriolis/tests/test_wsman.py

```diff
@@ -0,0 +1,157 @@
+    def test_connect_no_auth(self, mock_protocol):
+        self.conn.connect('url', 'username')
+        mock_protocol.assert_called_once_with(
+            endpoint='url',
+            transport='plaintext',
+            username='username',
+            password=None,
+            cert_pem=None,
+            cert_key_pem=None
+        )
+
+    @mock.patch.object(wsman.WSManConnection, 'connect')
+    @mock.patch('coriolis.utils.wait_for_port_connectivity')
+    def test_from_connection_info(self, mock_wait_for_port_connectivity,
+                                  mock_connect):
+        connection_info = {
+            "ip": "127.0.0.1",
+            "username": "user",
+            "password": "pass",
+        }
+        result = self.conn.from_connection_info(connection_info)
+        mock_wait_for_port_connectivity.assert_called_once_with(
+            "127.0.0.1", 5986)
+        mock_connect.assert_called_once_with(
+            url="https://127.0.0.1:5986/wsman", username="user",
+            password="pass", cert_pem=None, cert_key_pem=None)
+        self.assertIsInstance(result, self.conn.__class__)
+
+    def test_from_connection_info_missing_keys(self):
+        self.assertRaises(ValueError, self.conn.from_connection_info,
+                          {"username": "user", "password": "pass"})
+
+    def test_from_connection_info_invalid_type(self):
+        self.assertRaises(ValueError, self.conn.from_connection_info,
+                          'invalid-connection-type')
+
+    def test_disconnect(self):
+        self.conn.disconnect()
+        self.assertIsNone(self.conn._protocol)
+
+    def test_set_timeout(self):
+        self.conn.set_timeout(self.conn._conn_timeout)
+        self.assertEqual(self.conn._protocol.transport.timeout,
+                         self.conn._conn_timeout)
+
+    def test__exec_command(self):
+        self.conn._protocol.get_command_output.return_value = (
+            "std_out", "std_err", 0)
+        std_out, std_err, exit_code = self.conn._exec_command(
+            self.cmd, self.args)
+        self.assertEqual(std_out, "std_out")
+        self.assertEqual(std_err, "std_err")
+        self.assertEqual(exit_code, 0)
+
+        self.conn._protocol.open_shell.assert_called_once_with(
+            codepage=wsman.CODEPAGE_UTF8)
+        shell_id = self.conn._protocol.open_shell.return_value
+        self.conn._protocol.run_command.assert_called_once_with(
+            shell_id, self.cmd, self.args)
+        command_id = self.conn._protocol.run_command.return_value
+        self.conn._protocol.get_command_output.assert_called_once_with(
+            shell_id, command_id)
+
+    def test__exec_command_exception(self):
+        self.conn._protocol.get_command_output.side_effect = requests.\
+            exceptions.ReadTimeout
+        self.assertRaises(exception.OSMorphingWinRMOperationTimeout,
+                          self.conn._exec_command, self.cmd, self.args)
```

Comment:
Same as above. Even when exceptions occur, `cleanup_command` and `close_shell` should be called.